README.md
# Blinkbox::CommonConfig

Put your reference properties in `config/reference.properties` and any environmental stuff which makes sense for a development environment in `config/application.properties`, then load them and access them with:

```ruby
# $ gem install blinkbox-common_config
require "blinkbox/common_config"

properties = Blinkbox::CommonConfig.new

# Accessible with symbols
properties[:'rabbitmq.url']
# Or strings
properties["rabbitmq.url"]
```

It will also load the properties file referenced in the environment variable `CONFIG_URL` so long as the file exists, or the http(s) URI responds with a 200 status code.

CHANGELOG.md
# Changelog

## Open Source release (2015-01-28 14:11:21)

Today we have decided to publish parts of our codebase on Github under the [MIT licence](LICENCE). The licence is very permissive, but we will always appreciate hearing if and where our work is used!

## 0.4.0 ([#6](https://git.mobcastdev.com/Platform/common_config.rb/pull/6) 2014-10-03 15:41:55)

No deep setting

### Breaking change

- Should not have changed this to using deep setting (other apps depend on it being single keys, and this is more flexible)

## 0.3.1 ([#3](https://git.mobcastdev.com/Platform/common_config.rb/pull/3) 2014-08-18 09:46:37)

Depth

### Clarification

- Made improvements to the tests and code so that the form of nested properties when using the `tree` method is well defined.

## 0.3.0 ([#4](https://git.mobcastdev.com/Platform/common_config.rb/pull/4) 2014-09-25 10:00:48)

Symbols and booleans

### New feature

- Converts `"true"` and `"false"` to boolean values.
- Converts `":thing"` to `:thing` symbols.

## 0.2.0 ([#2](https://git.mobcastdev.com/Platform/common_config.rb/pull/2) 2014-08-13 16:33:04)

Use units in properties

### New Feature

- Properties in the format `5 seconds` will be converted to [Unit](https://github.com/olbrich/ruby-units) objects so that they can be easily manipulated:

```ruby
property = Unit("6 minutes")

p property.convert_to("seconds").scalar
### => 360
```

## 0.1.0

Initial release

### New features

- Reads `config/reference.properties`, overwrites with `config/application.properties` (if it exists) then overwrites with contents of the file or URL at `ENV['CONFIG_URL']` (if it exists/responds with HTTP 200)

spec/spec_helper.rb
$LOAD_PATH.unshift File.join(__dir__, "../lib")
require "blinkbox/common_config"

module Helpers
  def load_properties(properties_text)
    Dir.mktmpdir { |dir|
      open(File.join(dir, "reference.properties"), "w") do |f|
        f.puts properties_text
      end

      Blinkbox::CommonConfig.new(config_dir: dir)
    }
  end
end

RSpec.configure do |c|
  c.include Helpers
end

spec/common_config_spec.rb
context Blinkbox::CommonConfig do
  subject(:properties) {
    properties = load_properties <<-PROPS
      my_key = value
      another_key = another value
      logging.host = hostname.com
      logging.port = 1234
      logging = isn't in the tree for logging.
      delay = 5 seconds
      long_delay = 5 minutes
      nested.keys.depth = three
      nested.keys.other = three
      nested.other_key = two
      yes = true
      stringtrue = "true"
      no = false
      stringfalse = "false"
      symbol = :symbol
      something = "something"
    PROPS
  }

  describe "#[]" do
    it "must return the value alphanumeric property symbol keys" do
      expect(properties[:my_key]).to eq("value")
    end

    it "must be indifferent to symbol or string access" do
      expect(properties["my_key"]).to eq(properties[:my_key])
    end

    it "must return nil if the property is not defined" do
      expect(properties[:doesnt_exist]).to be_nil
    end

    it "must return a number with a unit for properties in the right format" do
      expect(properties[:delay].convert_to("seconds").scalar).to eq(5)
    end

    it "must return numbers with other units in a convertible format" do
      expect(properties[:long_delay].convert_to("seconds").scalar).to eq(5 * 60)
    end

    it "must return a 'true' value as a true boolean" do
      expect(properties[:yes]).to eql(true)
    end

    it "must return a 'false' value as a false boolean" do
      expect(properties[:no]).to eql(false)
    end

    it "must return '\"true\"' as the string 'true'" do
      expect(properties[:stringtrue]).to eql("true")
    end

    it "must return '\"false\"' as the string 'false'" do
      expect(properties[:stringtrue]).to eql("true")
    end

    it "must return '\"something\"' as the string 'something'" do
      expect(properties[:something]).to eql("something")
    end

    it "must return a symbol for strings starting with colons" do
      expect(properties[:symbol]).to eql(:symbol)
    end
  end

  describe "#tree" do
    it "must return all keys in the specified tree as a hash" do
      expected_hash = {
        host: "hostname.com",
        port: 1234
      }
      expect(properties.tree(:logging)).to eq(expected_hash)
    end

    it "must also return trees when using strings" do
      expect(properties.tree("logging")).to eq(properties.tree(:logging))
    end

    it "must return an empty hash if there are no matches" do
      expect(properties.tree(:absent)).to eq({})
    end

    it "must not return partial matches" do
      expect(properties.tree(:log)).to eq({})
    end

    it "must return nested keys as hashes" do
      expected_hash = {
        :'keys.depth' => "three",
        :'keys.other' => "three",
        :other_key    => "two"
      }
      expect(properties.tree(:nested)).to eq(expected_hash)
    end
  end
end

lib/blinkbox/common_config.rb
require "java_properties"
require "java_properties/data_types"
require "net/http"
require "tempfile"

module Blinkbox
  class CommonConfig
    # List of sources where properties have been loaded, with the most important first
    attr_reader :sources

    # Generates a CommonConfig object assuming that there is a `reference.properties` in the `config_fir` specified. If
    # an `application.properties` is also present it will load that too (overwriting any repeated property keys) and finally
    # if the environment variable `CONFIG_URL` is set it will load the properties from that (overwriting the others) whether
    # the URL is an absolute, relative (to the working directory, not the `config_dir`) or an HTTP URI.
    #
    # @param [String] config_dir The directory where a `reference.properties` exists and optionally an `application.conf` exists. Relative paths are taken from the working directory.
    # @param [#info, NilClass] logger A logger to which information about what properties have been loaded will be sent. Must respond to #info, can also be `nil`.
    def initialize(config_dir: "config", logger: nil)
      raise ArgumentError, "The logger given doesn't respond to #info." unless logger.nil? || logger.respond_to?(:info)

      reference_prop_file = File.expand_path(File.join(config_dir, "reference.properties"))
      application_prop_file = File.expand_path(File.join(config_dir, "application.properties"))
      raise RuntimeError, "No reference file at #{reference_prop_file}" unless File.exist?(reference_prop_file)

      @options = JavaProperties::Properties.new(reference_prop_file)
      logger.info "Loaded configuration from #{reference_prop_file}" unless logger.nil?
      @sources = [reference_prop_file]

      if File.exist?(application_prop_file)
        @options.load(application_prop_file)
        logger.info "Loaded configuration from #{application_prop_file}" unless logger.nil?
        @sources.unshift(application_prop_file)
      end

      if ENV['CONFIG_URL']
        if ENV['CONFIG_URL'] =~ %r{^https?://}
          res = Net::HTTP.get_response(URI.parse(ENV['CONFIG_URL']))
          raise "The CONFIG_URL points to a resource that returns an HTTP Status code of #{res.code}, not a 200." unless res.code == "200"

          remote_props = Tempfile.new('remote_properties')
          begin
            remote_props.write(res.body)
            remote_props.close
            @options.load(remote_props)
          ensure
            remote_props.unlink
          end
        else
          @options.load(ENV['CONFIG_URL'])
        end
        logger.info "Loaded configuration from #{ENV['CONFIG_URL']}" unless logger.nil?
        @sources.unshift(ENV['CONFIG_URL'])
      end
    end

    # Creates a CommonConfig object from a hash.
    #
    # @params [Hash] hash The hash which will become the config. Symbol or string key values are cool.
    def self.from_hash(hash)
      config = self.allocate
      config.instance_variable_set(:'@options', hash)
      config
    end

    # Accessor for the properties stored in the instance. Accepts strings or symbols indifferently.
    #
    # @params [String, Symbol] key The name of the property to retrieve.
    def [](key)
      @options[key]
    end
    alias :get :[]

    # Retrieves a hash of all properties which are beneath this starting element in the tree.
    #
    # # logging.udp.host = "127.0.0.1"
    # # logging.udp.port = 1234
    # # logging = this won't be returned
    #
    # properties.tree(:logging)
    # # => { :'udp.host' => "127.0.0.1", :'udp.port' => 1234 }
    #
    # properties.tree(:log)
    # # => {}
    #
    # @params [String, Symbol] root The root key to look underneath.
    # @returns [Hash] Returns a has
    def tree(root)
      hash = {}
      @options.each { |key, value|
        len = root.length + 1
        if key.to_s.slice(0, len) == root.to_s + '.'
          hash[key.to_s[len..-1].to_sym] = value
        end
      }
      hash
    end
  end
end